File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(arr, predicate)
  if #arr == 0 then -- Check if the array is empty
      return nil
  end

  for _, value in ipairs(arr) do -- Find first element that satisfies predicate
      if predicate(value) then
          return string.lower(value)
      end
  end

  return nil -- Nil if no element satisfies predicate
end

-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.create(function() -- Create the coroutine
      local power = 1 -- Initialize power
      while power <= limit do 
          coroutine.yield(power) 
          power = power * base -- Yeilds power until limit is reached
      end
  end)
end

-- Write your say function here
function say(word)
  local words = {} -- Empty table to store words
  
  if word then
      table.insert(words, word)
  else
    return ""
  end -- Store argument in table or return "" if no words passed

  return function(next_word)
      if next_word then
          table.insert(words, next_word)
          return say(table.concat(words, " ")) -- Continue the chain
      else
          return table.concat(words, " ") -- End the chain and return the sentence
      end
  end
end
  
-- Write your line count function here
function meaningful_line_count(filename)
  local file, err = io.open(filename, "r")
  if not file then
      error("No such file" .. err) -- Throws error
  end

  local count = 0
  for line in file:lines() do
      if line:match("%S") and not line:match("^%s*#") then -- Check if line is non-empty, not whitespace only, and doesn't start with '#'
          count = count + 1
      end
  end

  file:close()
  return count
end

-- Write your Quaternion table here
Quaternion = {}
Quaternion.__index = Quaternion

-- Constructor
function Quaternion.new(a, b, c, d)
    return setmetatable({a = a, b = b, c = c, d = d}, Quaternion)
end

-- Coefficients
function Quaternion:coefficients()
    return {self.a, self.b, self.c, self.d}
end

-- Addition
function Quaternion.__add(q1, q2)
    return Quaternion.new(
        q1.a + q2.a,
        q1.b + q2.b,
        q1.c + q2.c,
        q1.d + q2.d
    )
end

-- Multiplication
function Quaternion.__mul(q1, q2)
    return Quaternion.new(
        q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,
        q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,
        q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,
        q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
    )
end

-- Conjugate
function Quaternion:conjugate()
    return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

-- String representation
function Quaternion:__tostring()
    local str = ""
    if self.a ~= 0 then str = str .. tostring(self.a) end

    if self.b ~= 0 then
        if self.b > 0 and str ~= "" then str = str .. "+" end
        if self.b == -1 then
            str = str .. "-i"
        elseif self.b == 1 then
            str = str .. "i"
        else
            str = str .. tostring(self.b) .. "i"
        end
    end

    if self.c ~= 0 then
        if self.c > 0 and str ~= "" then str = str .. "+" end
        if self.c == -1 then
            str = str .. "-j"
        elseif self.c == 1 then
            str = str .. "j"
        else
            str = str .. tostring(self.c) .. "j"
        end
    end

    if self.d ~= 0 then
        if self.d > 0 and str ~= "" then str = str .. "+" end
        if self.d == -1 then
            str = str .. "-k"
        elseif self.d == 1 then
            str = str .. "k"
        else
            str = str .. tostring(self.d) .. "k"
        end
    end

    if str == "" then str = "0" end
    return str
end

-- Equality
function Quaternion.__eq(q1, q2)
    return q1.a == q2.a and q1.b == q2.b and q1.c == q2.c and q1.d == q2.d
end

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import Optional


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(a: str, p: bool, /) -> Optional[str]:
    for string in a:
        if p(string):
            return string.lower()
    return None

# Write your powers generator here
def powers_generator(*, base, limit) -> int:
    exponent = 0
    while base ** exponent <= limit:
        yield base ** exponent
        exponent += 1

# Write your say function here
def say(word = None, /) -> str:
    if word == None:
        return ""

    words = []

    def chain(next_word = None):
        if next_word is not None:
            words.append(next_word)
            return chain
        else:
            return ' '.join(words)

    if word is not None:
        words.append(word)

    return chain

# Write your line count function here
def meaningful_line_count(filename, /) -> int:
    count = 0
    try:
        with open(filename, "r") as f:
            for x in f:
                x_trunc = x.strip()
                if x_trunc and x_trunc[0] != "#":
                    count += 1
    except FileNotFoundError:
        raise FileNotFoundError('No such file')

    return count

# Write your Quarternion class here
@dataclass(frozen = True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float
    
    @property
    def coefficients(self, /) -> tuple:
        return (self.a, self.b, self.c, self.d)

    @property
    def conjugate(self, /) -> "Quaternion":
        return Quaternion(self.a, -self.b, -self.c, -self.d)
    
    def __add__(self, other: "Quaternion", /) -> "Quaternion":
        return Quaternion(
            self.a + other.a, 
            self.b + other.b, 
            self.c + other.c, 
            self.d + other.d)

    def __mul__(self, other: "Quaternion", /) -> "Quaternion":
        return Quaternion(
            self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
            self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
            self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
            self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        )

    def __eq__(self, other: "Quaternion", /) -> "Quaternion":
        return (self.a == other.a) and (self.b == other.b) and (self.c == other.c) and (self.d == other.d)

    def __str__(self, /) -> str:
        terms = []
        if self.a != 0:
            terms.append(f"{self.a}")
        if self.b != 0:
            if self.b == 1:
                terms.append("i")
            elif self.b == -1:
                terms.append("-i")
            else:
                terms.append(f"{self.b}i")
        if self.c != 0:
            if self.c == 1:
                terms.append("j")
            elif self.c == -1:
                terms.append("-j")
            else:
                terms.append(f"{self.c}j")
        if self.d != 0:
            if self.d == 1:
                terms.append("k")
            elif self.d == -1:
                terms.append("-k")
            else:
                terms.append(f"{self.d}k")

        if not terms:
            return "0"

        result = terms[0]
        for term in terms[1:]:
            if term[0] == "-":
                result += term
            else:
                result += f"+{term}"

        if result[0] == "+":
            result = result[1:]
        
        return result